source/TaskQueue.cpp

```diff
@@ -0,0 +1,207 @@
+/* TaskQueue.cpp
+Copyright (c) 2022 by Michael Zahniser
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "TaskQueue.h"
+
+#include <algorithm>
+#include <chrono>
+#include <condition_variable>
+#include <exception>
+
+using namespace std;
+
+namespace {
+
+	// The main task queue used by the worker threads.
+	queue<TaskQueue::Task> tasks;
+	mutex asyncMutex;
+	condition_variable asyncCondition;
+	bool shouldQuit = false;
+
+	// The maximum amount of sync tasks to execute in one go.
+	constexpr int MAX_SYNC_TASKS = 100;
+
+	// Worker threads for executing tasks.
+	struct WorkerThreads {
+		WorkerThreads() noexcept
+		{
+			threads.resize(max(4u, thread::hardware_concurrency()));
+			for(thread &t : threads)
+				t = thread(&TaskQueue::ThreadLoop);
+		}
+		~WorkerThreads()
+		{
+			{
+				lock_guard<mutex> lock(asyncMutex);
+				shouldQuit = true;
+			}
+			asyncCondition.notify_all();
+			for(thread &t : threads)
+				t.join();
+		}
+
+		vector<thread> threads;
+	} threads;
+}
+
+
+
+TaskQueue::~TaskQueue()
```

Comment:
this destructor makes the destructor of engine unneccessary

Reply:
I'd rather keep the destructor of Engine because if I remove it then the order where the data member TaskQueue is defined starts to matter,  or else there can be a race between the running task and the destruction of a data member. I thought it was less error prone to simply keep the destructor.

I could have also removed it and then moved the TaskQueue data member to the end of Engine with a comment explaining why, but idk imo the destructor is less dangerous so to speak.